### PR TITLE
Fixed expandable content does not work when using `enableVirtualization` in `Table`

### PR DESCRIPTION
**Pull request by smmr-dn**
## Changes

When enabling virtualization using the `virtualizer` from `@tanstack/react-virtual`, expandable content was reported broken since it was originally rendered with each original row, which was completely ignored by the `virtualizer`. To fix this problem, the expandable content in this PR is moved to a separate component and regenerated its index. The virtualizer now is made aware of the changes in the page size when sub-components are defined. These components are rendered accordingly to their indices associated with which row they are expanded from.

## Testing

Confirmed that sub-components are rendered correctly when `enableVirtualization` is set to true.
virtulization_after

An e2e test is added for expandable contents in virtualized table.

## Docs

Added both `patch` changesets for the bug fix and the removal of `WithCSSTransition` in sub-components.

## After-PR TODOs:
- [ ] Add a warning when user passes both sub-row and sub-component according to [comment](https://github.com/iTwin/iTwinUI/pull/2208#discussion_r1765409011) and [issue](https://github.com/iTwin/iTwinUI/issues/1750).
- [ ] Fix `WithCSSTransition` not working in sub-components ([comment](https://github.com/iTwin/iTwinUI/pull/2208#discussion_r1765394017))

### CODE REVIEW

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (mayank99) [inline packages/itwinui-react/src/core/Table/Table.tsx:601]: I don't think we can add random properties to the user's row, because these could conflict with the user's data.

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (mayank99) [inline packages/itwinui-react/src/core/Table/Table.tsx:835]: leftover

File: packages/itwinui-react/src/core/Table/TableExpandableRow.tsx
Comment (mayank99) [inline packages/itwinui-react/src/core/Table/TableExpandableRow.tsx:27]: nit: this extra variable looks unnecessary, the props can be directly set on the `Box` instead

File: packages/itwinui-react/src/core/Table/TableExpandableRow.tsx
Comment (mayank99) [inline packages/itwinui-react/src/core/Table/TableExpandableRow.tsx:11]: We should `memo`ize this component too.

Also I think some of the props can be removed/reorganized. For example `subComponent` could just be passed through `children`, and `tableRowRef` and `virtualizer` could be handled using a normal `forwardRef` pattern.

File: packages/itwinui-react/src/core/Table/TableExpandableRow.tsx
Comment (mayank99) [inline packages/itwinui-react/src/core/Table/TableExpandableRow.tsx:1]: an e2e test for virtualized + expandable content would be useful

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.tsx:595]: I'm not quite certain about this. What preference should I give use in `getSubRows`?

Comment (mayank99) [reply packages/itwinui-react/src/core/Table/Table.tsx:595]: I just mean that we should prefer the user's `getSubRows` prop over the one we have defined ourselves.

Think of our own `getSubRows` function as the "default" value that should only be used if the user doesn't pass a `getSubRows` prop to `Table`.

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.tsx]: I see, you're right. Users might have data with the same properties. How should I make the loop to stop in this case?

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.tsx]: I have had this removed.

File: packages/itwinui-react/src/core/Table/TableExpandableRow.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/TableExpandableRow.tsx]: I have set this directly on the `Box`.

File: packages/itwinui-react/src/core/Table/TableExpandableRow.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/TableExpandableRow.tsx:1]: Noted and will add tests asap.

File: packages/itwinui-react/src/core/Table/TableExpandableRow.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/TableExpandableRow.tsx]: I'm not quite sure how to combine the 2 refs before passing into the child component as forwarded ref so I passed the tableRef then merged them together inside the child. Is this the correct way to do this?

I have memoized the component.

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.tsx]: I have taken preference the user-defined `getSubRows`.

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (mayank99) [inline packages/itwinui-react/src/core/Table/Table.tsx:1]: The Table crashes when selecting a row.

https://github.com/user-attachments/assets/69a87370-2418-429e-838c-fe36331278f4

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.tsx]: To what I understand, getSubRows will get called when until there are no subRows left, which means it reaches the last "leaf" of the parent row. This implementation instead creates a new sub-row using the original row's data so technically it becomes its own sub-row, then gets called again as a row ready to become its own sub-row again. I added `hasParent` as a flag to mark that we already create a sub-row to be used as our sub-component. In order words, as soon as one child is created, the loop must break. 

When rendering the rows in `getPreparedRow`, when expanding row1, the expanded content is actually itself and the expanded content. The `isSubComponent` flag is there to mark that this additional row is actually a sub-component disguised as a sub-row so there is no need to render this row and just render the sub-component.
image

File: packages/itwinui-react/src/core/Table/TableExpandableRowMemoized.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/TableExpandableRowMemoized.tsx]: I have moved theses props outside.

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.tsx]: I have replaced these with `row`.

File: .changeset/lucky-suits-decide.md
Comment (smmr-dn) [reply .changeset/lucky-suits-decide.md]: I have edited the changeset.

File: packages/itwinui-react/src/core/Table/TableExpandableRowMemoized.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/TableExpandableRowMemoized.tsx:25]: I have double-checked and it's needed since without this style content might shrink or grow out of the flex container when there isn't enough space. I have put it back to the style.

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.tsx]: I have changed the approach so these are no longer needed.

File: packages/itwinui-react/src/core/Table/TableExpandableRow.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/TableExpandableRow.tsx:1]: I have added e2e test.

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.tsx:1]: This has been confirmed to be working for the new approach.

**Review by mayank99 — COMMENTED** [review]
@smmr-dn Can you please explain the new changes? I'm a bit lost, because this PR looks completely different from when I first reviewed it and I don't see any context anywhere around why these large changes were made.

**Comment by smmr-dn** [comment]
> @smmr-dn Can you please explain the new changes? I'm a bit lost, because this PR looks completely different from when I first reviewed it and I don't see any context anywhere around why these large changes were made.

For the new approach, instead of making the sub-component a sub-row, the sub-component now is a component with index. The virtualizer is also notified about this change in the page size when there are any expanded content/row on the screen.

- The `listOfRowsAndSubComponents` is an array of strings indicating the rows and currently expanded rows in the page. The length of this array is fed to the virtualizer so it will be notified every time the page changes its size due to expansion and collapse.

- The `getRowIndices` is an array which calculates the index of the subcomponent, whose data is associated with the expanded row. We use this index calculation to determine if the current row is the main row or subcomponent to render accordingly. (The list of virtual items only appends more rows and keeps the main rows in the original order, so that's why we need to map subcomponents with the correct indices).
> For example, if the row0 (index 0) is expanded then the index of the subcomponent0 is 1, row1 (index 2) is not expanded, row2 (index 3) is associated with subcomponent2 with the index of 4. Therefore, subcomponent2 should take the data from virtual item of index 2 and rendered as item of index 4. 

As long as we assign the right index to the correct subcomponent, the table should be rendered correctly.

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.tsx:805]: Should a different data structure, rather than an array, be used in this case?

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.tsx]: I have moved these into a custom hook called `useSubComponentVirtualizing`.

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.tsx]: Can I conditionally return within the same function?

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.tsx]: Since the `getRowIndexFromVirtualizerIndex` does the index mapping for us, can I make use of that returned index?

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.tsx]: I have done accordingly in my latest commit.

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.tsx]: I have made the change.

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.tsx]: I have done accordingly in my latest commit.

**Review by smmr-dn — COMMENTED** [review]
After some discussions, we have come to decide that we should go with the first approach, where we make the sub-component a sub-row and conditionally render the main row and sub-component. This PR has been updated accordingly to the approach.

File: testing/e2e/playwright.config.ts
Comment (smmr-dn) [inline testing/e2e/playwright.config.ts:8]: Since the test row number is 100000, it took a while to load the table and the sub-components so I boosted the timeout to 100s for it to work locally. This might be too much so if there is any other better solution, let me know!

Comment (mayank99) [reply testing/e2e/playwright.config.ts:8]: Can you describe the problem that led you to increase the timeout?

Is there a performance issue that we need to fix? I see in the playground that you're only testing it with a few rows. Could you try to testing it manually with 10_000 or 100_000?

Comment (smmr-dn) [reply testing/e2e/playwright.config.ts:8]: I have tested manually with 10,000 and 100,000 and the playground loaded pretty quickly. However, for the playwright test, I kept getting the timeout error until I increased the timeout to 100s, when the test finally passed locally.

File: packages/itwinui-react/src/core/Table/Table.test.tsx
Comment (mayank99) [inline packages/itwinui-react/src/core/Table/Table.test.tsx:2043]: Why was this changed? I would expect sub-rows to work the same as before.

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (mayank99) [inline packages/itwinui-react/src/core/Table/Table.tsx:587]: Since we are adding rows to this set and never removing them, this could grow indefinitely in size.

Consider using [`WeakSet`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakSet) instead, so that unused values are automatically garbage collected.

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (mayank99) [inline packages/itwinui-react/src/core/Table/Table.tsx:608]: nit: Is this the same as a regular ternary?

```suggestion
      getSubRows: subComponent ? getSubRowsWithSubComponents : getSubRows,
```

File: testing/e2e/playwright.config.ts
Comment (mayank99) [reply testing/e2e/playwright.config.ts]: Interestingly, it passes on my machine without the increased timeout. The UI is ready within a few hundred milliseconds. I understand that my M1 MacBook Pro might be more powerful than the average Windows laptop, but 100 seconds seems excessive.

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.tsx:863]: image

Right, it doesn't work with `WithCSSTransition`, should I add a changeset saying that there will no longer be transition animation in sub-components?

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.tsx:608]: I have added this as an after-PR TODO.

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.tsx:880]: You're right, it's always true. I have removed the condition.

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.tsx:608]: I have made it a regular ternary.

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.tsx:863]: I'm not entirely sure but I have noticed one thing is the `in` props is being `undefined` as we pass in `row.isExpanded` which was true when the sub-component was still in the main row. However, since sub-component is now a sub-row, `row.isExpanded` is no longer useful as we return the sub-component instead. `Row` in sub-component does not represent the main that can expand and collapse anymore.

File: packages/itwinui-react/src/core/Table/Table.test.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.test.tsx:2043]: I should have addressed and asked about this in the previous PR. This was how it originally worked before the separating sub-component PR and I was planning to fix this issue after attempting to transform the sub-component to be sub-row (as the previous PR was the continuation of this one). After I got the sub-component to work as a sub-row, it went back to the original behavior. I'm assuming it's because how we conditionally render the main row/sub-row but I'm not entirely sure about this.

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (mayank99) [reply packages/itwinui-react/src/core/Table/Table.tsx:863]: Let's try to fix this in a separate PR.

File: packages/itwinui-react/src/core/Table/Table.test.tsx
Comment (mayank99) [reply packages/itwinui-react/src/core/Table/Table.test.tsx:2043]: Ah ok, that makes sense. So it's the same as it was before #2240.

**Review by mayank99 — APPROVED** [review]
LGTM!

File: packages/itwinui-react/src/core/Table/Table.tsx
Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.tsx:863]: I have added this as an after-PR TODO.

Comment (smmr-dn) [reply packages/itwinui-react/src/core/Table/Table.tsx:863]: I have added the changeset.

File: .changeset/short-planets-listen.md
Comment (smmr-dn) [reply .changeset/short-planets-listen.md]: Changeset is updated.

Comment (mayank99) [reply .changeset/short-planets-listen.md]: I have to disagree about this needing a `minor`. I've said it before - we need to be very mindful of what goes in the `minor` section (usually API changes and significant behavioral changes). This change is not large enough to warrant a note in the `minor` section. In fact, the main bug fix is a bigger change the removal of the animation.

But I'll defer to your decision, I don't want to block this PR.

Comment (mayank99) [reply .changeset/short-planets-listen.md]: We can keep both as `patch` 👍

Comment (smmr-dn) [reply .changeset/short-planets-listen.md]: I have made them both `patch`

**Pull request merged**
